### PR TITLE
(7/9) Convert Memo protocol buffer

### PR DESCRIPTION
**Pull request by amiecorso**
## High Level Overview of Change
Provide a native Java wrapper for the [Memo protocol buffer](https://github.com/ripple/rippled/blob/develop/src/ripple/proto/org/xrpl/rpc/v1/transaction.proto#L80).

Analogue in Swift: https://github.com/xpring-eng/XpringKit/pull/136
Analogue in JS: https://github.com/xpring-eng/Xpring-JS/pull/255

### Context of Change
We want to provide an API which returns transaction history. This API shouldn't return raw protocol buffers as they have some slightly complex abstractions and because that means we're tying a network layer to a public API.

This is the start of building up several types to ultimately serve all transaction types.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Before / After
No change.

## Test Plan
New CI introduced for unit tests.

8 & 9 / building up to Transaction wrapper

### CODE REVIEW

**Comment by amiecorso** [comment]
The Circle failure currently due to outdated Testnet txn hash for integration tests.  Will open separate PR to address from master.

File: src/main/java/io/xpring/xrpl/model/XRPMemo.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        } else {
+            type = memo.getMemoType().getValue().toByteArray();
+        }
+        return XRPMemo.builder()
+                    .data(data)
+                    .format(format)
+                    .type(type)
+                    .build();
+    }
+}
```
Comment (keefertaylor) [inline src/main/java/io/xpring/xrpl/model/XRPMemo.java]: nit: newline

File: src/main/java/io/xpring/xrpl/model/XRPMemo.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Nullable byte[] format();
+
+    /**
+     * @return Hex value representing characters allowed in URLs.
+     * Conventionally, a unique relation (according to RFC 5988) that defines the format of this memo.
+     */
+    @Nullable byte[] type();
+
+    static XRPMemo from(Memo memo) {
+        byte[] data;
```
Comment (keefertaylor) [inline src/main/java/io/xpring/xrpl/model/XRPMemo.java]: I'd build these up one at a time to make it easier to read. Also, the null value is implicitly assigned, though you can make it explicit. 

```
byte[] data; // Or if you want to be super clear: byte[] data= null; 
if (memo.getMemoData().getValue() != ByteString.EMPTY) {
  data = memo.getMemoData().getValue().toByteArray();
}
        // Intentional blank line; let reader know we're done with mucking around with `data`
byte [] format; 
// and so on
```

If you really wanted to be succinct there, you could just use a ternary operator and do the assignment in two statements:
```
ByteString memoData = memo.getMemoData().getValue();
byte[] data = memoData == ByteString.EMPTY ? null : memoData.toByteArray()
        // Intentional blank line; let reader know we're done with mucking around with `data`
ByteString format = ...
```

I tend to prefer the second though some people find ternary operators hard to read.

File: src/main/java/io/xpring/xrpl/model/XRPMemo.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        byte[] format;
+        byte[] type;
+        if (memo.getMemoData().getValue() == ByteString.EMPTY) {
```
Comment (keefertaylor) [inline src/main/java/io/xpring/xrpl/model/XRPMemo.java]: I'm surprised that `==` WAI. `==` in java should compare object addresses and I'm surprised an arbitrary `EMPTY` ByteString has the same object address as another `EMPTY` byte string. 

We should probably use `.equals()` here:P
```
if (memo.getMemoData.getValue().equals(ByteString.EMPTY) {
```

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 4]
```diff
@@ -27,6 +27,32 @@
...
     static int testDestinationTag = 2;
+
+    static ByteString memoDataBytes;
+    static ByteString memoFormatBytes;
+    static ByteString memoTypeBytes;
+
+    static {
+        try {
+            memoDataBytes = ByteString.copyFrom("123", "Utf8");
+        } catch (UnsupportedEncodingException e) {
```
Comment (keefertaylor) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: don't use one letter variable names. Saves time now, difficult to read later :) 

(Yes, I know it's only test code but it's good to have good habits)

File: src/main/java/io/xpring/xrpl/model/XRPMemo.java (hunk 2 above)
Comment (amiecorso) [reply src/main/java/io/xpring/xrpl/model/XRPMemo.java]: wayyyyyy prettier thank you

File: src/main/java/io/xpring/xrpl/model/XRPMemo.java (hunk 3 above)
Comment (amiecorso) [reply src/main/java/io/xpring/xrpl/model/XRPMemo.java]: I think that the reason this is WAI is that when you construct a default, empty `MemoData` object, its bytestring value is explicitly set to that constant `ByteString.EMPTY` in the constructor. `.equals` sounds good to me and is also easier to read in the ternary operator

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java (hunk 4 above)
Comment (amiecorso) [reply src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: unlearning bad habits they taught me at school

**Comment by amiecorso** [comment]
CircleCI seems to be crashing for reasons unrelated to tests...

**Comment by amiecorso** [comment]
\+ Noah cause Dino's out

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 5]
```diff
@@ -16,8 +16,8 @@
     static {
         try {
             testCurrencyCode = ByteString.copyFrom("123", "Utf8");
-        } catch (UnsupportedEncodingException e) {
-            e.printStackTrace();
+        } catch (UnsupportedEncodingException error) {
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: Again, calling this `e` is a very normal thing in java, but if you must use a full word, let's call this `exception`

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 6]
```diff
@@ -16,8 +16,8 @@
...
+            error.printStackTrace();
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: Let's use a logger here.  Look at `DefaultIlpClient` on how to set up a logger

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 7]
```diff
@@ -26,6 +26,32 @@
...
+
+    static ByteString memoDataBytes;
+    static ByteString memoFormatBytes;
+    static ByteString memoTypeBytes;
+
+    static {
+        try {
+            memoDataBytes = ByteString.copyFrom("123", "Utf8");
+        } catch (UnsupportedEncodingException error) {
+            error.printStackTrace();
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: logger and `exception` here as well

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 8]
```diff
@@ -26,6 +26,32 @@
...
+        }
+    }
+    static {
+        try {
+            memoFormatBytes = ByteString.copyFrom("456", "Utf8");
+        } catch (UnsupportedEncodingException error) {
+            error.printStackTrace();
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: same same

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 9]
```diff
@@ -26,6 +26,32 @@
...
+        }
+    }
+    static {
+        try {
+            memoTypeBytes = ByteString.copyFrom("789", "Utf8");
+        } catch (UnsupportedEncodingException error) {
+            error.printStackTrace();
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java]: same same

**Comment by amiecorso** [comment]
If you don't mind checking out my logger!  I saw how to set one up but couldn't find an example of it in-use so hopefully what I've got here is ok.

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 10]
```diff
@@ -16,8 +20,8 @@
     static {
         try {
             testCurrencyCode = ByteString.copyFrom("123", "Utf8");
-        } catch (UnsupportedEncodingException e) {
-            e.printStackTrace();
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create testCurrencyCode", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:24]: This should be `logger.error` or `logger.warn`

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 11]
```diff
@@ -26,6 +30,32 @@
...
+            memoFormatBytes = ByteString.copyFrom("456", "Utf8");
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create memoFormatBytes", exception);
+        }
+    }
+    static {
+        try {
+            memoTypeBytes = ByteString.copyFrom("789", "Utf8");
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create memoTypeBytes", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:56]: same here

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 12]
```diff
@@ -26,6 +30,32 @@
...
+            memoDataBytes = ByteString.copyFrom("123", "Utf8");
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create memoDataBytes", exception);
+        }
+    }
+    static {
+        try {
+            memoFormatBytes = ByteString.copyFrom("456", "Utf8");
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create memoFormatBytes", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:49]: same here

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 13]
```diff
@@ -26,6 +30,32 @@
...
+
+    static ByteString memoDataBytes;
+    static ByteString memoFormatBytes;
+    static ByteString memoTypeBytes;
+
+    static {
+        try {
+            memoDataBytes = ByteString.copyFrom("123", "Utf8");
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create memoDataBytes", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:42]: same

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 14]
```diff
@@ -37,17 +67,17 @@
     static {
         try {
             testPublicKey = ByteString.copyFrom("123", "Utf8");
-        } catch (UnsupportedEncodingException e) {
-            e.printStackTrace();
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create testPublicKey", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:71]: same

File: src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java [hunk 15]
```diff
@@ -37,17 +67,17 @@
...
     }
 
     static ByteString testTransactionSignature;
     static {
         try {
             testTransactionSignature = ByteString.copyFrom("456", "Utf8");
-        } catch (UnsupportedEncodingException e) {
-            e.printStackTrace();
+        } catch (UnsupportedEncodingException exception) {
+            logger.info("Can't create testTransactionSignature", exception);
```
Comment (nkramer44) [inline src/test/java/io/xpring/xrpl/FakeXRPProtobufs.java:80]: same

**Comment by amiecorso** [comment]
great, all updated to logger.error!

**Pull request merged**